sdk/documentintelligence/azure-ai-documentintelligence/src/main/java/com/azure/ai/documentintelligence/implementation/SyncOperationLocationPollingStrategy.java

```diff
@@ -0,0 +1,234 @@
+        T value;
+        if (data == null) {
+            value = null;
+        } else if (type.getJavaClass().isAssignableFrom(BinaryData.class)) {
+            value = type.getJavaClass().cast(data.toReplayableBinaryData());
+        } else {
+            value = data.toObject(type, serializer);
+        }
+        return value;
+    }
+
+    private static final HttpHeaderName RETRY_AFTER_MS_HEADER = HttpHeaderName.fromString("retry-after-ms");
+    private static final HttpHeaderName X_MS_RETRY_AFTER_MS_HEADER = HttpHeaderName.fromString("x-ms-retry-after-ms");
+    private static Duration getRetryAfterFromHeaders(HttpHeaders headers, Supplier<OffsetDateTime> nowSupplier) {
+        // Found 'x-ms-retry-after-ms' header, use a Duration of milliseconds based on the value.
+        Duration retryDelay = tryGetRetryDelay(headers, X_MS_RETRY_AFTER_MS_HEADER, s -> tryGetDelayMillis(s));
+        if (retryDelay != null) {
+            return retryDelay;
+        }
+
+        // Found 'retry-after-ms' header, use a Duration of milliseconds based on the value.
+        retryDelay = tryGetRetryDelay(headers, RETRY_AFTER_MS_HEADER, s -> tryGetDelayMillis(s));
+        if (retryDelay != null) {
+            return retryDelay;
+        }
+
+        // Found 'Retry-After' header. First, attempt to resolve it as a Duration of seconds. If that fails, then
+        // attempt to resolve it as an HTTP date (RFC1123).
+        retryDelay = tryGetRetryDelay(headers, HttpHeaderName.RETRY_AFTER,
+            headerValue -> tryParseLongOrDateTime(headerValue, nowSupplier));
+
+        // Either the retry delay will have been found or it'll be null, null indicates no retry after.
+        return retryDelay;
+    }
+    private static Duration tryGetRetryDelay(HttpHeaders headers, HttpHeaderName headerName,
+                                             Function<String, Duration> delayParser) {
+        String headerValue = headers.getValue(headerName);
+
+        return CoreUtils.isNullOrEmpty(headerValue) ? null : delayParser.apply(headerValue);
+    }
+    private static Duration tryParseLongOrDateTime(String value, Supplier<OffsetDateTime> nowSupplier) {
+        long delaySeconds;
+        try {
+            OffsetDateTime retryAfter = new DateTimeRfc1123(value).getDateTime();
+
+            delaySeconds = nowSupplier.get().until(retryAfter, ChronoUnit.SECONDS);
+        } catch (DateTimeException ex) {
+            delaySeconds = tryParseLong(value);
+        }
+
+        return (delaySeconds >= 0) ? Duration.ofSeconds(delaySeconds) : null;
+    }
+    private static long tryParseLong(String value) {
+        try {
+            return Long.parseLong(value);
+        } catch (NumberFormatException ex) {
+            return -1;
+        }
+    }
+    private static Duration tryGetDelayMillis(String value) {
+        long delayMillis = tryParseLong(value);
+        return (delayMillis >= 0) ? Duration.ofMillis(delayMillis) : null;
+    }
```

Comment:
getRetryDelay is the most complex code here, though Sync/Async should share the same code.